tests/integration/s2n_test_constants.py

```diff
@@ -282,7 +296,46 @@
         client_sni="www.rattlesnake.com",
         client_ciphers="ECDHE-RSA-AES128-SHA",
         expected_cert=SNI_CERTS["quail_cn_rattlesnake_cn"],
-        expect_matching_hostname=False)]
+        expect_matching_hostname=False),
+    MultiCertTest(
+        description="Test cert with embedded wildcard is not treated as a wildcard.",
+        server_certs=[ SNI_CERTS["alligator"], SNI_CERTS["embedded_wildcard"] ],
+        client_sni="www.labelstartWILDCARDlabelend.com",
+        client_ciphers="ECDHE-RSA-AES128-SHA",
+        expected_cert=SNI_CERTS["alligator"],
+        expect_matching_hostname=False),
+    MultiCertTest(
+        description="Test non empty left label wildcard cert is not treated as a wildcard."\
+                    " s2n only supports wildcards with a single * as the left label",
+        server_certs=[ SNI_CERTS["alligator"], SNI_CERTS["non_empty_label_wildcard"] ],
+        client_sni="WILDCARD.middle.end",
+        client_ciphers="ECDHE-RSA-AES128-SHA",
+        expected_cert=SNI_CERTS["alligator"],
+        expect_matching_hostname=False),
+    MultiCertTest(
+        description="Test cert with trailing * is not treated as wildcard.",
+        server_certs=[ SNI_CERTS["alligator"], SNI_CERTS["trailing_wildcard"] ],
+        client_sni="the.prefix.WILDCARD",
+        client_ciphers="ECDHE-RSA-AES128-SHA",
+        expected_cert=SNI_CERTS["alligator"],
+        expect_matching_hostname=False),
+    MultiCertTest(
+        description="Certificate with exact sni match(termite.insect.hexapod) is preferred over wildcard"\
+                    " *.insect.hexapod",
+        server_certs=[ SNI_CERTS["wildcard_insect"], SNI_CERTS["alligator"], SNI_CERTS["termite"] ],
+        client_sni="termite.insect.hexapod",
+        client_ciphers="ECDHE-RSA-AES128-SHA",
+        expected_cert=SNI_CERTS["termite"],
+        expect_matching_hostname=True)]
+# Positive test for wildcard matches
+MULTI_CERT_TEST_CASES.extend([MultiCertTest(
+        description="Test wildcard *.insect.hexapod matches subdomain " + specific_insect_domain,
+        server_certs=[ SNI_CERTS["alligator"], SNI_CERTS["wildcard_insect"] ],
+        client_sni=specific_insect_domain,
+        client_ciphers="ECDHE-RSA-AES128-SHA",
```

Comment:
Is there test coverage for the cases that combine key type and wildcard/exact? For example, wildcard ECDSA with an exact RSA for different cipher suite preferences?

Reply:
There isn't. I'll add one that verifies the exact match is preferred even when there is only a match for a lower key type preference:

```
cert_matches: certA_exact_rsa, certB_wc_ecdsa
server_ciphers: ECDSA-*:RSA-*
client_ciphers: ECDSA-*:RSA-*
expected cert: certA_exact_rsa
```

Reply:
I think 
e015b8c  addresses this